smartsim/_core/_cli/cli.py

```diff
@@ -28,70 +28,84 @@
 
 import argparse
 import sys
-
-from smartsim._core._cli.build import Build
-from smartsim._core._cli.clean import Clean
-from smartsim._core._cli.utils import get_install_path
-
-
-def _usage() -> str:
-    usage = [
-        "smart <command> [<args>]\n",
-        "Commands:",
-        "\tbuild       Build SmartSim dependencies (Redis, RedisAI, ML runtimes)",
-        "\tclean       Remove previous ML runtime installation",
-        "\tclobber     Remove all previous dependency installations",
-        "\nDeveloper:",
-        "\tsite        Print the installation site of SmartSim",
-        "\tdbcli       Print the path to the redis-cli binary" "\n\n",
-    ]
-    return "\n".join(usage)
-
+import typing as t
+
+# from pkg_resources import require
+
+# import smartsim._core._cli as cli
+from smartsim._core._cli.utils import MenuItemConfig
+from smartsim._core._cli.clean import (
+    configure_parser as clean_parser,
+    execute as clean_execute,
+    execute_all as clobber_execute,
+)
+import smartsim._core._cli.clean as clean
+from smartsim._core._cli.site import execute as site_execute
+from smartsim._core._cli.dbcli import execute as dbcli_execute
+from smartsim._core._cli.build import configure_parser as build_parser, execute as build_execute
 
 class SmartCli:
-    def __init__(self) -> None:
+    def __init__(self, menu: t.List[MenuItemConfig]) -> None:
+        self.menu: t.Dict[str, MenuItemConfig] = {item.command: item for item in menu}
         parser = argparse.ArgumentParser(
-            description="SmartSim command line interface", usage=_usage()
+            prog="smart",
+            description="SmartSim command line interface",
         )
-
-        parser.add_argument("command", help="Subcommand to run")
-
-        # smart
-        if len(sys.argv) < 2:
-            parser.print_help()
-            sys.exit(0)
-
-        args = parser.parse_args(sys.argv[1:2])
-        if not hasattr(self, args.command):
-            parser.print_help()
-            sys.exit(0)
-        getattr(self, args.command)()
-
-    def build(self) -> None:
-        Build()
-        sys.exit(0)
-
-    def clean(self) -> None:
-        Clean()
-        sys.exit(0)
-
-    def clobber(self) -> None:
-        Clean(clean_all=True)
-        sys.exit(0)
-
-    def site(self) -> None:
-        print(get_install_path())
-        sys.exit(0)
-
-    def dbcli(self) -> None:
-        bin_path = get_install_path() / "_core" / "bin"
-        for option in bin_path.iterdir():
-            if option.name in ("redis-cli", "keydb-cli"):
-                print(option)
-                sys.exit(0)
-        print("Database (Redis or KeyDB) dependencies not found")
-        sys.exit(1)
-
-
-def main() -> None:
-    SmartCli()
+        self.parser = parser
+        self.args: t.Optional[argparse.Namespace] = None
```

Comment:
Purely stylistic: If we are going to have the `self.parsed_args` propery, can we make `self.args` into "protected" `self._args`, or do we need to keep public write access to that attr for some reason?

Reply:
100%. victim of my own iteration.